tools/iree-e2e-matmul-test.cc

```diff
@@ -0,0 +1,1247 @@
+               results->n, n_start, n_end, results->result_type,
+               results->actual_contents.data, results->expected_contents.data,
+               emoji(false));
+  fprintf(file, "\n");
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_make_status(IREE_STATUS_ABORTED);
+}
+
+// Helper for check_matmul_results: the actual interesting part once we've
+// obtained and validated the {m,k,n}_size values. On error, detailed logging is
+// written to |file| if it is not NULL.
+static iree_status_t check_matmul_results_impl(FILE* file,
+                                               const matmul_results_t* results,
+                                               int check_every) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  int count = 0;
+  for (iree_hal_dim_t m = 0; m < results->m; ++m) {
+    for (iree_hal_dim_t n = 0; n < results->n; ++n) {
+      if (++count < check_every) continue;
+      count = 0;
+      iree_e2e_test_value_t actual_value =
+          read_matrix_element(results->m, results->n, results->result_type,
+                              results->actual_contents.data, m, n);
+      iree_e2e_test_value_t expected_value =
+          read_matrix_element(results->m, results->n, results->result_type,
+                              results->expected_contents.data, m, n);
+      if (!matmul_result_elements_agree(actual_value, expected_value)) {
+        iree_status_t status = check_matmul_failure(
+            file, results, actual_value, expected_value, m, n, check_every);
+        IREE_TRACE_ZONE_END(z0);
+        return status;
+      }
+    }
+  }
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+// Given an actual matmul's inputs and output (all host-local), uses a reference
+// matmul implementation on the same inputs to check if the output is correct.
+// On error, detailed logging is written to |file| if it is not NULL.
+static iree_status_t check_matmul_results(FILE* file,
+                                          const matmul_results_t* results) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  int check_every = calculate_check_every(results->m, results->n);
+  iree_status_t status = check_matmul_results_impl(file, results, check_every);
+  if (!iree_status_is_ok(status) && check_every > 1) {
+    // If we got a failure with check_every>1, that didn't log a useful
+    // numerical summary, as most of the reference matrix entries hadn't been
+    // computed. Rerun now with check_every=1 to get that numerical logging.
+    iree_status_ignore(status);
+    status = check_matmul_results_impl(file, results, 1);
+  }
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+//===----------------------------------------------------------------------===//
+// RNG utilities
+//===----------------------------------------------------------------------===//
+
+// Parameter for locally defined lcg similar to std::minstd_rand.
+#define IREE_PRNG_MULTIPLIER 48271
+#define IREE_PRNG_MODULUS 2147483647
```

Comment:
Now that this is C++, you can use `std::minstd_rand` rather than reimplementing it.
```c++
std::minstd_rand prng;
int value = prng();  // calls std::minstd_rand::operator()
```
That's deterministic and implementation-independent. What would be implementation-dependent would be using `std::*_distribution` and `std::default_random_engine`.

Reply:
I'd rather keep it simple and C - it's only C++ for the custom module helper and I'd like to replace that in the future and that'll be harder if we start sprinkling in C++.